### PR TITLE
add Result.flatMap and Result.join

### PR DESCRIPTION
**Pull request by kingsleyh**
Hey - I don't know if you find these useful addition - but I'm using them on my project so thought I would create a PR for you in case you want to include them in the core - if not it's cool I can keep them in an Result.Extra

### CODE REVIEW

File: source/Result.mint [hunk 1]
```diff
@@ -42,4 +42,17 @@ module Result {
     })()
     `
   }
+
+  fun join (input : Result(b, Result(a, b))) : Result(a, c) {
```
Comment (gdotdesign) [inline source/Result.mint]: Shouldn't this be `fun join (input: Result(error, Result(error, value))) : Result(error, value)`?

File: source/Result.mint [hunk 2]
```diff
@@ -42,4 +42,17 @@ module Result {
...
+    if (Result.isOk(input)) {
+      `input.value`
+    } else {
+      input
+    }
+  }
+
+  fun flatMap (func : Function(b, Result(a, c)), input : Result(a, b)) : Result(a, c) {
```
Comment (gdotdesign) [inline source/Result.mint]: Shouldn't this be `fun flatMap (input: Function(value, Result(error, newValue)) : Result(error, newValue)`?

**Review by gdotdesign — COMMENTED** [review]
I think these functions can be useful, the type parameters needs some clean up, also I'll probably rewrite them to use descriptive names instead of `a` and `b`

**Comment by kingsleyh** [comment]
I followed the same pattern that Elm uses - so for join:

```
{-| Flattens nested Maybes
    join (Just (Just 1)) == Just 1
    join (Just Nothing)  == Nothing
    join Nothing         == Nothing
-}
join : Maybe (Maybe a) -> Maybe a
join mx =
    case mx of
        Just x ->
            x

        Nothing ->
            Nothing
```

and for flatMap:

```
flatMap : (a -> Maybe b) -> Maybe a -> Maybe b
flatMap f maybe =
  Maybe.map f maybe
    |> join
```

I tried using the lines you suggested but then the tests fail - unless you think that the tests are not doing the right thing also?

**Comment by kingsleyh** [comment]
I've tidied up the type params and fixed Join

**Comment by kingsleyh** [comment]
There are lots of possible implementations of flatMap - another possibility which also passes the tests are this one:
```
fun flatMap (func : Function(a, Result(error, b)), input : Result(error, a)) : Result(error, b) {
    if (Result.isOk(input)) {
      `func(input.value)`
    } else {
      `new Err()`
    }
  }
```

**Comment by gdotdesign** [comment]
Thank you! It's fine now :+1: :slightly_smiling_face:

**Pull request merged**